### PR TITLE
fix port sg_ids

### PR DESCRIPTION
**Pull request by morozovalekseywot**
Add new argument for show diffs on external changes in security groups

### CODE REVIEW

**Review by ftersin — CHANGES_REQUESTED** [review]
add a change log

File: vkcs/networking/resource_port.go [hunk 1]
```diff
@@ -105,6 +105,14 @@ func ResourceNetworkingPort() *schema.Resource {
 				Description: "(Conflicts with `security_group_ids`) If set to `true`, then no security groups are applied to the port. If set to `false` and no `security_group_ids` are specified, then the port will yield to the default behavior of the Networking service, which is to usually apply the \"default\" security group.",
 			},
 
+			"show_security_group_ids_diff": {
```
Comment (ftersin) [inline vkcs/networking/resource_port.go:108]: It brings full control of port's security groups. Not only viewing diff.

File: vkcs/networking/resource_port.go [hunk 2]
```diff
@@ -105,6 +105,14 @@ func ResourceNetworkingPort() *schema.Resource {
 				Description: "(Conflicts with `security_group_ids`) If set to `true`, then no security groups are applied to the port. If set to `false` and no `security_group_ids` are specified, then the port will yield to the default behavior of the Networking service, which is to usually apply the \"default\" security group.",
 			},
 
+			"control_security_group_ids": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				ForceNew:    false,
+				Default:     false,
+				Description: "If this argument is set to true, then the user has full control over `security_group_ids`.",
```
Comment (ftersin) [inline vkcs/networking/resource_port.go:113]: Not user but thr resource. Also we need to specify behavior w/ default group.

Comment (ftersin) [reply vkcs/networking/resource_port.go:113]: ... and difference between empty set and no specified argument

Comment (morozovalekseywot) [reply vkcs/networking/resource_port.go]: change to full_security_group_control

File: vkcs/networking/resource_port.go [hunk 3]
```diff
@@ -105,6 +105,14 @@ func ResourceNetworkingPort() *schema.Resource {
 				Description: "(Conflicts with `security_group_ids`) If set to `true`, then no security groups are applied to the port. If set to `false` and no `security_group_ids` are specified, then the port will yield to the default behavior of the Networking service, which is to usually apply the \"default\" security group.",
 			},
 
+			"control_security_group_ids": {
```
[the remark this answers is not in the document]
Comment (morozovalekseywot) [reply vkcs/networking/resource_port.go]: change to full_security_group_control

File: examples/networking/port/main.tf [hunk 4]
```diff
@@ -14,5 +14,7 @@ resource "vkcs_networking_port" "persistent_etcd" {
   }
   # Specify required security groups instead of getting 'default' one
   security_group_ids = [vkcs_networking_secgroup.etcd.id]
-  tags               = ["tf-example", "etcd"]
+  # Set this argument to true, if you want to have full control over security_group_ids
```
Comment (ftersin) [inline examples/networking/port/main.tf:17]: Set this argument to true every time

File: examples/networking/port/main.tf [hunk 5]
```diff
@@ -14,5 +14,7 @@ resource "vkcs_networking_port" "persistent_etcd" {
...
+  full_security_groups_control = true
+  tags                       = ["tf-example", "etcd"]
```
Comment (ftersin) [inline examples/networking/port/main.tf:19]: reformat

File: vkcs/networking/resource_port.go [hunk 6]
```diff
@@ -95,16 +95,26 @@ func ResourceNetworkingPort() *schema.Resource {
 				ForceNew:    false,
 				Elem:        &schema.Schema{Type: schema.TypeString},
 				Set:         schema.HashString,
-				Description: "(Conflicts with `no_security_groups`) A list of security group IDs to apply to the port. The security groups must be specified by ID and not name (as opposed to how they are configured with the Compute Instance).",
+				Computed:    true,
+				Description: "(Conflicts with `no_security_groups`) A list of security group IDs to apply to the port. The security groups must be specified by ID and not name. External changes are not displayed by default, to see them you need to set the `full_security_groups_control` argument to true.",
```
Comment (ftersin) [inline vkcs/networking/resource_port.go:99]: could we use schema feature for conflict detection?

Comment (ftersin) [inline vkcs/networking/resource_port.go:99]: A list of security group IDs to apply to the port. The security groups must be specified by ID and not name. If the list is empty no one security group is attached to the port. If the argument is absent `vkcs_networking_port` resource does not control security groups of the port just reads them. _note_ This behavior is actual with `full_security_groups_control` = true only and introduced in X.Y.Z version of the provider. Legacy behavior is still supported but deprecated and too confusing to be described.

File: vkcs/networking/resource_port.go [hunk 7]
```diff
@@ -402,6 +412,9 @@ func resourceNetworkingPortRead(ctx context.Context, d *schema.ResourceData, met
 	// This can be different from what the user specified since
 	// the port can have the "default" group automatically applied.
 	d.Set("all_security_group_ids", port.SecurityGroups)
+	if d.Get("full_security_groups_control").(bool) {
+		d.Set("security_group_ids", port.SecurityGroups)
+	}
```
Comment (ftersin) [inline vkcs/networking/resource_port.go:417]: Could we write a deprecation message if full_security_groups_control has false value?

File: examples/networking/port/main.tf (hunk 4 above)
Comment (morozovalekseywot) [reply examples/networking/port/main.tf]: done

File: examples/networking/port/main.tf (hunk 5 above)
Comment (morozovalekseywot) [reply examples/networking/port/main.tf]: done

File: vkcs/networking/resource_port.go (hunk 6 above)
Comment (morozovalekseywot) [reply vkcs/networking/resource_port.go]: no( We need a case with sg_ids = [] and no_sgs = true

Comment (morozovalekseywot) [reply vkcs/networking/resource_port.go]: done

File: vkcs/networking/resource_port.go [hunk 8]
```diff
@@ -95,16 +95,26 @@ func ResourceNetworkingPort() *schema.Resource {
...
+				Deprecated:  "Configure `full_security_groups_control = true` and `security_group_ids = []` instead",
 				Description: "(Conflicts with `security_group_ids`) If set to `true`, then no security groups are applied to the port. If set to `false` and no `security_group_ids` are specified, then the port will yield to the default behavior of the Networking service, which is to usually apply the \"default\" security group.",
 			},
 
+			"full_security_groups_control": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				ForceNew:    false,
+				Default:     false,
+				Description: "If this argument is set to true, then the terraform has full control over `security_group_ids`. So, if you create a port with `full_security_groups_control = true` and `security_group_ids = []`, the port will not have a default security group.",
```
[the remark this answers is not in the document]
Comment (morozovalekseywot) [reply vkcs/networking/resource_port.go]: done

File: vkcs/networking/resource_port.go (hunk 7 above)
Comment (morozovalekseywot) [reply vkcs/networking/resource_port.go:417]: add mesage

File: CHANGELOG.md [hunk 9]
```diff
@@ -7,6 +7,10 @@ description: |-
 
 # VKCS Provider's changelog
 
+### v0.7.5 (unreleased)
```
[the remark this answers is not in the document]
Comment (morozovalekseywot) [reply CHANGELOG.md]: bump

**Review by paaanic — APPROVED** [review]

**Review by ftersin — APPROVED** [review]

**Pull request merged**